File: DataStructures/DataStructures/LinkedList.m
//
// Created by Raj on 4/1/16.
// Copyright (c) 2016 Rajusa. All rights reserved.
//

#import "LinkedList.h"

@implementation LinkedListNode
@synthesize data = data_;
@synthesize next = next_;
@end

@implementation LinkedList

- (instancetype) init
{
    if (self = [super init])
    {
        head_ = nil;
        tail_ = nil;
    }

    return self;
}

- (void) dealloc
{
    [self removeAllObjects];
}

@synthesize count = count_;

- (void) addObject: (NSObject*) object
{
    LinkedListNode* newNode = [[LinkedListNode alloc] init];
    newNode.data = object;

    if (tail_)
        tail_.next = newNode;
    tail_ = newNode;

    if (!head_)
        head_ = newNode;

    count_++;
}

- (NSObject*) removeFirstObject
{
    if (!head_)
        return nil;

    LinkedListNode* node = head_;
    if (tail_ == head_)
        tail_ = nil;
    head_ = head_.next;

    count_--;
    return node.data;
}

- (NSObject*) removeLastObject
{
    if (!head_)
        return nil;

    LinkedListNode* current = head_;
    if (current == tail_)
    {
        head_ = nil;
        tail_ = nil;
        count_--;
        return current.data;
    }

    while (current != nil)
    {
        if (current.next == tail_)
        {
            LinkedListNode* lastNode = current.next;
            current.next = nil;
            tail_ = current;
            count_--;
            return lastNode.data;
        }
        current = current.next;
    }

    return nil;
}

- (NSObject*) removeObject: (NSObject*) object
{
    if (!head_)
        return nil;

    LinkedListNode* previous = nil;
    LinkedListNode* current = head_;
    while (current != nil)
    {
        if (current.data == object)
        {
            if (head_ == current)
                head_ = current.next;

            if (tail_ == current)
                tail_ = previous;

            if (previous)
                previous.next = current.next;

            count_--;
            return current.data;
        }

        previous = current;
        current = current.next;
    }

    return nil;
}

- (void) removeAllObjects
{
    if (!head_)
        return;

    LinkedListNode* current = head_;
    while (current != nil)
    {
        LinkedListNode* next = current.next;
        current.data = nil;
        current.next = nil;
        current = next;
    }

    head_ = nil;
    tail_ = nil;
    count_ = 0;
}

- (BOOL) contains: (NSObject*) object
{
    if (!head_)
        return NO;

    LinkedListNode* current = head_;
    while (current != nil)
    {
        if (current.data == object)
            return YES;

        current = current.next;
    }

    return NO;
}

- (NSObject*) getFirstObject
{
    if (head_)
        return head_.data;

    return nil;
}

- (NSObject*) getLastObject
{
    if (tail_)
        return tail_.data;

    return nil;
}

- (NSObject*) getObjectAtIndex: (NSUInteger) index
{
    if (!head_)
        return nil;

    LinkedListNode* current = head_;
    NSUInteger position = 0;
    while (current != nil)
    {
        if (position == index)
            return current.data;

        position++;
        current = current.next;
    }
    return nil;
}

- (void) printAllObjectsWithDataType: (Class) dataType
{
    if (!head_)
        return;

    LinkedListNode* current = head_;
    while (current != nil)
    {
        if ([NSNumber class] == dataType)
        {
            NSNumber* object = (NSNumber*)current.data;
            NSLog(@"%@", [object stringValue]);
        }
        current = current.next;
    }
}
@end

File: DataStructures/DataStructures/LinkedList.swift
//
// Created by Raj on 2019-01-13.
// Copyright (c) 2019 Rajusa. All rights reserved.
//

import Foundation

class SwLinkedListNode
{
    var data: NSObject?;
    fileprivate var next: SwLinkedListNode?;

    init(data: NSObject) {
        self.data = data;
    }

    init(data: NSObject, next: SwLinkedListNode) {
        self.data = data;
        self.next = next;
    }
}

public class SwLinkedList : NSObject, DataCollection, DataCollectionDebug
{
    var head: SwLinkedListNode?;
    var tail: SwLinkedListNode?;
    var nodeCount: UInt = 0;

    public class func newInstance() -> SwLinkedList {
        return SwLinkedList()
    }

    public func count() -> UInt {
        return self.nodeCount;
    }

    public func add(_ object: NSObject!) {
        let newNode = SwLinkedListNode(data: object);

        if let tail = self.tail {
            tail.next = newNode;
        }

        self.tail = newNode;

        if self.head == nil {
            self.head = newNode;
        }

        self.nodeCount += 1;
    }

    public func removeFirstObject() -> NSObject? {
        guard let head = self.head else {
            return nil;
        }

        if let tail = self.tail {
            if tail === head {
                self.tail = nil;
            }
        }

        self.head = head.next;
        self.nodeCount -= 1;
        return head.data;
    }

    public func removeLastObject() -> NSObject? {
        guard let head = self.head else {
            return nil;
        }

        var current : SwLinkedListNode? = head;
        if let tail = self.tail {
            if tail === head {
                self.head = nil;
                self.tail = nil;
                self.nodeCount = 0;
                return current!.data;
            }
        }

        while current != nil {
            if (current!.next === self.tail) {
                let lastNode : SwLinkedListNode? = current!.next;
                current!.next = nil;
                self.tail = current;
                self.nodeCount -= 1;
                return lastNode!.data;
            }

            current = current!.next;
        }

        return nil;
    }

    public func remove(_ object: NSObject!) -> NSObject! {
        guard let head = self.head else {
            return nil;
        }

        var previous : SwLinkedListNode? = nil;
        var current : SwLinkedListNode? = head;
        while current != nil {
            if (current!.data == object)
            {
                if head === current {
                    self.head = current!.next;
                }

                if self.tail === current {
                    self.tail = previous;
                }

                if previous != nil {
                    previous!.next = current!.next;
                }
                return current!.data;
            }

            previous = current;
            current = current!.next;
        }
        return nil;
    }

    public func removeAllObjects()
    {
        guard let head = self.head else {
            return;
        }

        // Note: We need to go through and release one object at a time
        // else we get a stack overflow
        var current : SwLinkedListNode? = head;
        while current != nil {
            let next: SwLinkedListNode? = current!.next;
            current!.data = nil;
            current!.next = nil;
            current = next;
        }

        self.head = nil;
        self.tail = nil;
        self.nodeCount = 0;
    }

    public func contains(_ object: NSObject) -> Bool {
        guard let head = self.head else {
            return false;
        }

        var current : SwLinkedListNode? = head;
        while current != nil {
            if (current!.data == object) {
                return true;
            }

            current = current!.next;
        }
        return false;
    }

    public func getFirstObject() -> NSObject? {
        guard let head = self.head else {
            return nil;
        }

        return head.data;
    }

    public func getLastObject() -> NSObject? {
        guard let tail = self.tail else {
            return nil;
        }

        return tail.data;
    }

    public func getObjectAt(_ index: UInt) -> NSObject? {
        guard let head = self.head else {
            return nil;
        }

        var current : SwLinkedListNode? = head;
        var position : UInt = 0;
        while current != nil {
            if (position == index) {
                return current!.data;
            }

            position += 1;
            current = current!.next;
        }

        return nil;
    }

    public func printAllObjects(withDataType dataType: AnyClass) {
        guard let head = self.head else { return; }
        var current: SwLinkedListNode? = head;
        while current != nil {
            if current?.data.self is NSNumber {
                let data: NSNumber = current?.data as! NSNumber;
                NSLog("%@", data.stringValue);
            }

            current = current?.next;
        }
    }
}
